elegant/clean_timepoint_data.py

```diff
@@ -0,0 +1,147 @@
+        if not dry_run:
+            pos_md = [timepoint_data for timepoint_data in pos_md if timepoint_data['timepoints'] != timepoint]
+            datafile.json_encode_atomic_legible_to_file(pos_md, experiment_root / position / 'position_metadata.json')   # Write out new position_metadata
+
+    position_annotation_file = experiment_root / 'annotations' / f'{position}.pickle'
+    if position_annotation_file.exists():
+        general_annotations,timepoint_annotations = load_data.read_annotation_file(position_annotation_file)
+
+        if timepoint in timepoint_annotations:
+            print(f'Found entry for removal in annotation for position {position}')
+            if not dry_run:
+                del timepoint_annotations[timepoint]
+                load_data.write_annotation_file(position_annotation_file,
+                    general_annotations, timepoint_annotations)
+
+def remove_timepoint_from_experiment(experiment_root, timepoint, dry_run=False):
+    """Removes information about a timepoint from all positions in an experiment (as well as experiment_metadata)
+
+    This function is useful for when a data for an entire timepoint is totally unusable
+        (e.g. metering fails hard, one or more files get fatally overwritten).
+
+    Parameters
+        experiment_root - str/pathlib.Path pointing to experiment directory
+        timepoint - str with standard format for timepoint acquisition (YYYY-MM-DDtHHMM)
+        dry_run - bool flag that toggles taking action (if False, only specifies when offending files are found);
+            this flag is exposed for corresponding functionality through remove_timepoint_for_position
+
+    Example Usage
+        experiment_root = /path/to/experiment
+        timepoint = '2018-07-30t1200'
+        remove_timepoint_for_position(experiment_root, position, timepoint)
+    """
+
+    experiment_root = pathlib.Path(experiment_root)
+    positions = [metadata_file.parent for metadata_file in sorted(experiment_root.glob('*/position_metadata.json'))]
+
+    for position in positions:
+        remove_timepoint_for_position(experiment_root, position, timepoint, dry_run=dry_run)
+
+    md_file = (experiment_root/'experiment_metadata.json')
+    with md_file.open() as md_fp:
+        expt_md = json.load(md_fp)
+
+    try:
+        timepoint_idx = expt_md['timepoints'].index(timepoint)
+    except ValueError:
+        print('Timepoint not found in experiment_metadata')
+        return
+
+    for list_entry_type in ['durations','timestamps','timepoints']:
+        del expt_md[list_entry_type][timepoint_idx]
+
+    for dict_entry_type in ['brightfield metering', 'fluorescent metering', 'humidity', 'temperature']:
+        expt_md[dict_entry_type] = {key:val for key,val in expt_md[dict_entry_type].items() if key != timepoint}
+
+    datafile.json_encode_atomic_legible_to_file(expt_md,experiment_root/'experiment_metadata.json')
+
+def remove_excluded_positions(experiment_root,dry_run=False):
```

Comment:
This function doesn't remove the position entries in the `experiment_metadata.json` files, nor the annotation files. Should it? Having some but not all might cause trouble with downstream code that isn't scrupulous about error and consistency checking.

Reply:
I err on the side of keeping the data around (at least in the annotations). Even if one deletes a bunch of positions for having some class of bad worms (e.g. out-of-focus, bursting, larval arresting), it's useful to keep the records of those events around for review/reference. This might make more sense for the experiment_metadata, though it's equivocal. The only issue I see is having to make a note that the "brightfield_meter_position" may no longer exist....